### PR TITLE
feat: support signature file in verify cmd

### LINKED ISSUE #1067: cosign verify not working with signature file
**Description**

We have an option to output signature to the standard output with the cosign sign command, but there is no way to use this signature with verify command.

Shouldn’t we accept this signature while using verify command? Because my signature is not uploaded to the registry, so, I can’t verify it with verify command, right?

you can reproduce what I’m saying by issuing the following commands on your terminal:

```shell
$ cosign generate-key-pair
$ cosign sign --upload=false --key cosign.key <yourdockerid>/alpine:latest > my.sig
$ cosign verify --key cosign.pub <yourdockerid>/alpine:latest
# it should fail, we should pass the signature to the verify command to make it success
```

cc: @Dentrax

### PR DESCRIPTION
**Pull request by developer-guy**
Signed-off-by: Batuhan Apaydın <batuhan.apaydin@trendyol.com>
Co-authored-by: Furkan Türkal <furkan.turkal@trendyol.com>

#### Summary

Add ability to verify via signature reference in verify cmd

#### Ticket Link

Fixes #1067 

#### Release Note

```release-note
feat: support signature file in verify cmd
```

### CODE REVIEW

File: pkg/cosign/verify.go [hunk 1]
```diff
@@ -281,6 +337,11 @@ func VerifyImageSignatures(ctx context.Context, signedImgRef name.Reference, co
 	return checkedSignatures, bundleVerified, nil
 }
 
+func isb64(data []byte) bool {
```
Comment (naveensrinivasan) [inline pkg/cosign/verify.go:340]: A suggestion. Why not inline this in the main function? This reduces the readability from jumping from one function to another. In go it is generally not a practice to have a small function with a simple check.

**Comment by developer-guy** [comment]
kindly ping @dlorenc

**Comment by dlorenc** [comment]
looks like this one got hit with a merge conflict. It LGTM otherwise.

**Comment by developer-guy** [comment]
it should be okay now @dlorenc

File: pkg/cosign/verify.go [hunk 2]
```diff
@@ -196,15 +213,54 @@ func VerifyImageSignatures(ctx context.Context, signedImgRef name.Reference, co
...
-		return nil, false, err
+
+	var sigs oci.Signatures
+	sigRef := co.SignatureRef
+	if sigRef == "" {
+		sigs, err = se.Signatures()
+		if err != nil {
+			return nil, false, err
+		}
+	} else {
```
Comment (hectorj2f) [inline pkg/cosign/verify.go:228]: nit: I'd suggest moving this piece of code to a function to reduce the size and logic in this function.

File: pkg/cosign/verify.go [hunk 3]
```diff
@@ -230,7 +286,7 @@ func VerifyImageSignatures(ctx context.Context, signedImgRef name.Reference, co
 					return err
 				}
 				if cert == nil {
-					return errors.New("no certificate found on signature")
+					return errors.New("no certificate found on sigRef")
```
Comment (hectorj2f) [inline pkg/cosign/verify.go:289]: Suggestion: `signature` sounds correct to me, it refers to the command argument name, imho.

File: pkg/cosign/verify.go (hunk 1 above)
Comment (hectorj2f) [reply pkg/cosign/verify.go:340]: I agree with @naveensrinivasan, wrapping a function call inside of another function isn't ideal.

File: pkg/cosign/verify.go (hunk 2 above)
Comment (developer-guy) [reply pkg/cosign/verify.go:228]: thank you @hectorj2f, done.

File: pkg/cosign/verify.go (hunk 3 above)
Comment (developer-guy) [reply pkg/cosign/verify.go]: thank you @hectorj2f, done.

File: pkg/cosign/verify.go (hunk 1 above)
Comment (developer-guy) [reply pkg/cosign/verify.go]: thank you @hectorj2f, done.

**Review by hectorj2f — APPROVED** [review]
lgtm 👍🏻

**Pull request merged**